### PR TITLE
Fail if exported devices contain invalid fields

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Validate exported devices exported, and fail if any fields are invalid.

#### Changes

- Use `ttnpb.EndDevice.ValidateFields` to validate that the exported device does not contain any errors
- Fail if validation fails
- Take this chance to replace `_` occurences with `-`, since `_` was an allowed character in TTN V2.

#### Testing

Export a device with name `some_device`.

- Without this change: No errors, device exported with invalid id, importing to TTS fails.
- With this change: Device exported as `some-device` instead, and all fields validated.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

### CODE REVIEW

File: cmd/util.go [hunk 1]
```diff
@@ -34,16 +34,35 @@ func toJSON(dev *ttnpb.EndDevice) ([]byte, error) {
 var (
 	errExport = errors.Define("export", "export device `{device}`")
 	errFormat = errors.DefineCorruption("format", "format device `{device}`")
+
+	errFailedValidation = errors.DefineInvalidArgument("failed_validation", "failed to validate device")
```
Comment (johanstokking) [inline cmd/util.go:38]: Do not use "fail" or "failed" in errors. Errors indicate failure.

File: cmd/util.go [hunk 2]
```diff
@@ -34,16 +34,35 @@ func toJSON(dev *ttnpb.EndDevice) ([]byte, error) {
...
 )
 
+func replaceUnderscores(s string) string {
```
Comment (johanstokking) [inline cmd/util.go:41]: Replace with what?

Please be clear and not require people to look at the implementation, even if it's a one liner, to see what's happening.

You can also do a package-global var like:

```go
var sanitizeID = strings.NewReplacer("-", "_")
```

This allows more advanced implementations without touching callers, when we implement some custom `Replace()` method

File: cmd/util.go [hunk 3]
```diff
@@ -34,16 +34,35 @@ func toJSON(dev *ttnpb.EndDevice) ([]byte, error) {
...
+	if err := dev.ValidateFields(ttnpb.EndDeviceFieldPathsNested...); err != nil {
+		return errFailedValidation.WithAttributes(
+			"device_id", dev.DeviceID,
+			"dev_eui", dev.DevEUI,
+		).WithCause(err)
+	}
 	b, err := toJSON(dev)
 	if err != nil {
-		return errFormat.WithAttributes("device", devEui).WithCause(err)
+		return errFormat.WithAttributes(
```
Comment (johanstokking) [inline cmd/util.go:62]: Attributes don't match with error description

**Review by johanstokking — CHANGES_REQUESTED** [review]
Please be a bit less sloppy and look at the diff next time when requesting review

**Review by htdvisser — COMMENTED** [review]
Nothing to add

**Review by johanstokking — APPROVED** [review]

File: cmd/util.go [hunk 4]
```diff
@@ -32,18 +32,34 @@ func toJSON(dev *ttnpb.EndDevice) ([]byte, error) {
...
+	dev, err := s.ExportDevice(devID)
 	if err != nil {
-		return errExport.WithAttributes("device", devEui).WithCause(err)
+		return errExport.WithAttributes("device_id", devID).WithCause(err)
+	}
+	// V3 does not allow any underscores in identifiers
+	dev.DeviceID = sanitizeID.Replace(dev.DeviceID)
+	dev.ApplicationID = sanitizeID.Replace(dev.ApplicationID)
+
+	if err := dev.ValidateFields(ttnpb.EndDeviceFieldPathsNested...); err != nil {
```
Comment (rvolosatovs) [inline cmd/util.go:51]:
```suggestion
	if err := dev.ValidateFields(); err != nil {
```

**Pull request merged**